swift/common/middleware/versioned_writes/oio_object_versioning.py

```diff
@@ -0,0 +1,738 @@
+            ver_cont_req = make_pre_authed_request(
+                req.environ, path=wsgi_quote(versions_cont_path),
+                method='DELETE', headers=hdrs, swift_source='OV')
+
+            # TODO: what if this one fails??
+            resp = ver_cont_req.get_response(self.app)
+            drain_and_close(resp)
+
+        if self._response_headers is None:
+            self._response_headers = []
+        for key, val in self._response_headers:
+            if key.lower() == SYSMETA_VERSIONS_ENABLED:
+                self._response_headers.extend([
+                    (CLIENT_VERSIONS_ENABLED.title(), val)])
+
+        start_response(self._response_status,
+                       self._response_headers,
+                       self._response_exc_info)
+        return app_resp
+
+    def _list_versions(self, req, start_response, location):
+        # Only supports JSON listings
+        req.environ['swift.format_listing'] = False
+        if not req.accept.best_match(['application/json']):
+            raise HTTPNotAcceptable(request=req)
+
+        params = req.params
+        # FIXME(FVE) this is probably not working with oio-sds backend
+        if 'version_marker' in params:
+            if 'marker' not in params:
+                raise HTTPBadRequest('version_marker param requires marker')
+
+            if params['version_marker'] != 'null':
+                try:
+                    ts = Timestamp(params.pop('version_marker'))
+                except ValueError:
+                    raise HTTPBadRequest('invalid version_marker param')
+
+                params['marker'] = self._build_versions_object_name(
+                    params['marker'], ts)
+
+        delim = params.get('delimiter', '')
+        # Exclude the set of chars used in version_id from user delimiters
+        if set(delim).intersection('0123456789.%s' % RESERVED_STR):
+            raise HTTPBadRequest('invalid delimiter param')
+
+        null_listing = []
+        subdir_set = set()
+
+        account = req.split_path(3, 3, True)[1]
+        versions_req = make_pre_authed_request(
+            req.environ, method='GET', swift_source='OV',
+            path=wsgi_quote('/v1/%s/%s' % (account, location)),
+            headers={'X-Backend-Allow-Reserved-Names': 'true'},
+        )
+        versions_req.environ['oio.query'] = {'versions': True}
+
+        # NB: no end_marker support (yet)
+        versions_req.params = {
+            k: params.get(k, '')
+            for k in ('prefix', 'marker', 'limit', 'delimiter', 'reverse',
+                      'format')}
+        versions_resp = versions_req.get_response(self.app)
+
+        if versions_resp.status_int == HTTP_NOT_FOUND:
+            raise versions_resp
+        elif is_success(versions_resp.status_int):
+            try:
+                listing = json.loads(versions_resp.body)
+            except ValueError:
+                app_resp = [versions_resp.body]
```

Comment:
Well seen! We should fix it on `openstack/swift` and/or fix it in `object_versioning.py` as well.